BaseTools/Plugin/HostBasedUnitTestRunner/HostBasedUnitTestRunner.py

```diff
@@ -251,117 +256,91 @@ def gen_code_coverage_clang(self, thebuilder):
         for testFile in testList:
-            lcovFile = f"{testFile}.lcov"
-            ret = RunCmd("llvm-cov", f"export -format=lcov --instr-profile={mergedProfData} {testFile} > {lcovFile} ")
-            if ret != 0:
-                logging.error("UnitTest Coverage: Failed to generate coverage data for " + testFile)
+            if self.clang_gen_lcov_xml(mergedProfData, [testFile], f"{testFile}.coverage.lcov", f"{testFile}.coverage.xml") != 0:
                 return 1
-            coveragexmlFile = f"{testFile}.coverage.xml"
-            ret = RunCmd("lcov_cobertura",f"{lcovFile} -o {coveragexmlFile}")
-            if ret != 0:
-                logging.error("UnitTest Coverage: Failed generate filtered coverage XML.")
-                return 1
-        # Merge all XML files
-        ret = self.merge_cobertura_xml_files([f"{testFile}.coverage.xml" for testFile in testList], mergedCoverageXml)
-        if ret != 0:
-            logging.error("UnitTest Coverage: Failed to merge coverage XML files.")
+
+        # Generate LCOV and XML for the package
+        if self.clang_gen_lcov_xml(mergedProfData, testList, mergedCoverageLcov, mergedCoverageXml) != 0:
             return 1
+
+        allProfrawlistFile = os.path.join(workspace, 'Build', 'profrawlist.txt')
+        allMergedProfData = os.path.join(workspace, 'Build', 'merged.profdata')
+        allCoverageLcov = os.path.join(workspace, 'Build', 'coverage.lcov')
+        allCoverageXml = os.path.join(workspace, 'Build', 'coverage.xml')
+
+        for file in [allProfrawlistFile, allMergedProfData, allCoverageLcov, allCoverageXml]:
+            if os.path.isfile(file):
+                os.remove(file)
+
+        # Generate LCOV and XML for all packages
+        if self.clang_gen_profdata(workspace, allProfrawlistFile, allMergedProfData) != 0:
+            return 1
+        if self.clang_gen_lcov_xml(allMergedProfData, allTestList, allCoverageLcov, allCoverageXml) != 0:
+            return 1
+
         return 0
 
-    def merge_cobertura_xml_files(self, xml_file_list, output_file):
+    def clang_gen_profdata(self, basepath, output_profrawlistfile, output_profdata):
         """
-        Merge multiple Cobertura XML files into a single output file.
+        Merge multiple LLVM profraw files into a single profdata file.
+        """
+        if not os.path.isdir(basepath):
+            logging.error(f"clang_gen_profdata: Base path {basepath} is not a directory.")
+            return 1
+        if not output_profrawlistfile:
+            logging.error("clang_gen_profdata: No profraw list file provided.")
+            return 1
+        if not output_profdata:
+            logging.error("clang_gen_profdata: No output profdata file provided.")
+            return 1
 
-        Args:
-            xml_file_list (list): List of Cobertura XML file paths to merge.
-            output_file (str): Path to the merged output XML file.
+        with open(output_profrawlistfile, "w") as f:

```

Comment:
Should this have a check for no `.profraw` files being present and give an explicit error message and return `1`?

Like:

```python
    profraw_files = glob.glob(os.path.join(basepath, "**", "*.profraw"), recursive=True)
    if not profraw_files:
       ...
```

Reply:
That would force user to always does clean and build.
We can do delete first if it's present.